### PR TITLE
Dynamic Referencing of specs, test cases and test code

### PR DESCRIPTION
**Pull request by thebioengineer**
Create a method for dynamic referencing across specifications, test cases and test code.

This PR creates an R6 object that will persist within a validation build that will store the numbering for every spec and test case. The numbering right now is generated as it comes across the references, but we could add a method for the user to indicate the reference number at the beginning. 

Additionally, there is a function to implement the R6 object. In this scenario, we can just add this function into existing processing functions to be able to add dynamic referencing without impacting anything else.

### CODE REVIEW

File: R/dynamic_referencing.R [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+#' test code ('test_code'). This allows numbering to increase independently for each.
+#' However, every reference must be unique. ie. @@reference whether it shows up in a 
+#' test case or spec will be the same reference.
+#' 
+#' The method 'reference_insertion' takes a vector of strings and replaces references
+#' with their numeric values.
+#' 
+#' 
+#' @keywords internal
+#' @export vt_dynamic_referencer
```
Comment (elimillera) [inline R/dynamic_referencing.R:29]: Is this function going to be used by a user? Or is it being exported for documentation purposes?

**Review by elimillera — APPROVED** [review]
I love this! This was one of our major pain points when we were designing our validation docs. I'm not extremely familiar with R6 but I've tested this out a bit and its working as I expect.

File: R/dynamic_referencing.R (hunk 1 above)
Comment (thebioengineer) [reply R/dynamic_referencing.R:29]: so its a situation kind of like the reporter in testthat. Not really intended for most users to be interacting with it, but it needs to be exported

**Review by mariev — COMMENTED** [review]
building off my other comment about bookdown compatibility, using an R function to capture dynamic numbering requires every child doc to be processed multiple times.

That is, 
1. process raw rmd to add numbering
2. split (1) into (a) roxygen block (b) report content 
3. process (a) into summary tables 
4. import (b), executing test code if applicable

This assumes all validation report elements (requirements, test cases, test code) are inputs for further processing, and not child docs. It also places a burden on users if elements are intended to be shared with a latex-only workflow.

A child doc - whether bookdown or rmarkdown - is imported wholly, and {bookdown} specifically has some nice features for handling many child docs. Namely that the `_bookdown.yml` is structured so that the report can be built using syntax below:

```
book_filename: "validation_report_VX"
rmd_files: [
  "summary_table.Rmd",
   "requirement_1.Rmd",
   "testcase_1.Rmd",
   "testcode_1.Rmd"
  ]
output_dir: "docs"
```

I think there are definitely pros to using either paradigm, we just need to be explicit about what {valtools} will support.  Some examples captured in  https://github.com/phuse-org/valtools/blob/dynamic_referencing/tests/testthat/test-latex_dynamic_referencing.R

**Review by mariev — APPROVED** [review]
getting a couple devtools::test() errors but it looks to be from `develop` branch code. Moved outstanding concern to issue #28 so that we can start playing with code

**Pull request merged**